### ISSUE TITLE
twisted.internet.tcp.Server.repstr wrong when listening on port 0

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2826]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2826|
|Type|defect|
|Created|2007-09-08 22:59:55Z|

The 0 ends up in the string and sticks there, instead of the actual port number which gets bound.

<details><summary>Searchable metadata</summary>

```
trac-id__2826 2826
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1189292395000000 1189292395000000
changetime__1194290753000000 1194290753000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

(In [21478]) A test and a fix.

Refs [#2826](https://github.com/twisted/twisted/issues/2826)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve removed owner|
|-|-|

This is ready to review in server-repstr-2826.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @therve|
|-|-|

Since test_serverRepr uses addCleanup for connector.disconnect, I think it might as well use it for port.stopListening too, instead of using the kinda-brokenish PortCleanerUpper feature of appending to self.ports.  This'll make it easier to get rid of PortCleanerUpper later (for which there's a ticket, I think).

It'd probably be good to assert against the exact string expected from !__repr!__, too.

I think this has a cleanup error on Windows and probably on OS X sometimes, since it creates 3 selectables: the port, the server, and the client.  It only takes care of cleaning up two of them: the port and the client.  This will be okay if the OS delivers notification of disconnection to the server within the one iteration window which port.stopListening creates, but if it takes longer or port.stopListening stops doing callLater(0 then it'll be a problem.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

(In [21540]) Process review comments.

Refs [#2826](https://github.com/twisted/twisted/issues/2826)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve removed owner|
|-|-|

That should be better!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @therve|
|-|-|

```MyProtocolFactoryMixin.protocolConnectionMade``` is called back with the ```MyProtocol``` instance now - the documentation still says it's called back with ```None```.

All else, excellent.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve commented|
|-|-|

(In [21587]) Add a comment.

Refs [#2826](https://github.com/twisted/twisted/issues/2826)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve set status to `closed`|
|-|-|

(In [21588]) Merge server-repstr-2826

Author: therve
Reviewer: exarkun
Fixes [#2826](https://github.com/twisted/twisted/issues/2826)

Fix the representation of a tcp Server connection so that it contains the
correct port when listening on port 0.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
